File: java/fory-core/src/main/java/org/apache/fory/serializer/DecimalSerializer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.fory.serializer;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.math.MathContext;
import org.apache.fory.config.Config;
import org.apache.fory.context.ReadContext;
import org.apache.fory.context.WriteContext;
import org.apache.fory.exception.DeserializationException;
import org.apache.fory.memory.MemoryBuffer;

/** Serializer for {@link BigDecimal} in native and xlang modes. */
public final class DecimalSerializer extends ImmutableSerializer<BigDecimal> implements Shareable {
  static final int MAX_MAGNITUDE_BYTES = 10_000;
  static final int MAX_MAGNITUDE_BITS = MAX_MAGNITUDE_BYTES * Byte.SIZE;
  // Compare scale bounds directly because Math.abs(Integer.MIN_VALUE) overflows.
  private static final int MAX_SCALE = 10_000;
  private static final BigInteger LONG_MIN = BigInteger.valueOf(Long.MIN_VALUE);
  private static final BigInteger LONG_MAX = BigInteger.valueOf(Long.MAX_VALUE);
  private final boolean xlang;

  public DecimalSerializer(Config config) {
    super(config, BigDecimal.class);
    xlang = config.isXlang();
  }

  @Override
  public void write(WriteContext writeContext, BigDecimal value) {
    if (xlang) {
      writeXlang(writeContext, value);
    } else {
      writeNative(writeContext, value);
    }
  }

  @Override
  public BigDecimal read(ReadContext readContext) {
    if (xlang) {
      return readXlang(readContext);
    }
    return readNative(readContext);
  }

  private void writeNative(WriteContext writeContext, BigDecimal value) {
    MemoryBuffer buffer = writeContext.getBuffer();
    int scale = value.scale();
    if (scale < -MAX_SCALE || scale > MAX_SCALE) {
      throw new IllegalArgumentException("Decimal scale out of range: " + scale);
    }
    BigInteger unscaled = value.unscaledValue();
    if (magnitudeExceedsLimit(unscaled)) {
      throw new IllegalArgumentException(
          "Decimal magnitude exceeds " + MAX_MAGNITUDE_BYTES + " bytes");
    }
    byte[] bytes = unscaled.toByteArray();
    buffer.writeVarUInt32Small7(scale);
    buffer.writeVarUInt32Small7(value.precision());
    buffer.writeVarUInt32Small7(bytes.length);
    buffer.writeBytes(bytes);
  }

  private BigDecimal readNative(ReadContext readContext) {
    MemoryBuffer buffer = readContext.getBuffer();
    int scale = buffer.readVarUInt32Small7();
    if (scale < -MAX_SCALE || scale > MAX_SCALE) {
      throw new DeserializationException("Decimal scale out of range: " + scale);
    }
    int precision = buffer.readVarUInt32Small7();
    int len = buffer.readVarUInt32Small7();
    checkBinaryBodyLength(len);
    buffer.checkReadableBytes(len);
    if (len == MAX_MAGNITUDE_BYTES + 1 && magnitudeExceedsLimit(buffer, len)) {
      throw new DeserializationException(
          "Decimal magnitude exceeds " + MAX_MAGNITUDE_BYTES + " bytes");
    }
    byte[] bytes = buffer.readBytes(len);
    BigInteger bigInteger = new BigInteger(bytes);
    return new BigDecimal(bigInteger, scale, new MathContext(precision));
  }

  private void writeXlang(WriteContext writeContext, BigDecimal value) {
    writeXlangDecimal(writeContext.getBuffer(), value.scale(), value.unscaledValue());
  }

  private BigDecimal readXlang(ReadContext readContext) {
    return readXlangDecimal(readContext.getBuffer());
  }

  static void writeXlangDecimal(MemoryBuffer buffer, int scale, BigInteger unscaled) {
    if (scale < -MAX_SCALE || scale > MAX_SCALE) {
      throw new IllegalArgumentException("Decimal scale out of range: " + scale);
    }
    if (canUseSmallEncoding(unscaled)) {
      long smallValue = unscaled.longValue();
      long header = encodeZigZag64(smallValue) << 1;
      buffer.writeVarInt32(scale);
      buffer.writeVarUInt64(header);
      return;
    }

    if (magnitudeExceedsLimit(unscaled)) {
      throw new IllegalArgumentException(
          "Decimal magnitude exceeds " + MAX_MAGNITUDE_BYTES + " bytes");
    }
    int sign = unscaled.signum() < 0 ? 1 : 0;
    BigInteger abs = unscaled.abs();
    byte[] magnitudeBytes = toCanonicalLittleEndianMagnitude(abs);
    long meta = (((long) magnitudeBytes.length) << 1) | sign;
    long header = (meta << 1) | 1L;
    buffer.writeVarInt32(scale);
    buffer.writeVarUInt64(header);
    buffer.writeBytes(magnitudeBytes);
  }

  static BigDecimal readXlangDecimal(MemoryBuffer buffer) {
    int scale = buffer.readVarInt32();
    if (scale < -MAX_SCALE || scale > MAX_SCALE) {
      throw new IllegalArgumentException("Decimal scale out of range: " + scale);
    }
    return new BigDecimal(readXlangUnscaled(buffer), scale);
  }

  static BigInteger readXlangBigInteger(MemoryBuffer buffer) {
    int scale = buffer.readVarInt32();
    if (scale != 0) {
      throw new IllegalArgumentException(
          "Cannot deserialize xlang decimal with scale " + scale + " into BigInteger");
    }
    return readXlangUnscaled(buffer);
  }

  private static BigInteger readXlangUnscaled(MemoryBuffer buffer) {
    long header = buffer.readVarUInt64();
    if ((header & 1L) == 0L) {
      return BigInteger.valueOf(decodeZigZag64(header >>> 1));
    }
    long meta = header >>> 1;
    int sign = (int) (meta & 1L);
    long lenLong = meta >>> 1;
    if (lenLong <= 0 || lenLong > Integer.MAX_VALUE) {
      throw new IllegalArgumentException(
          "Invalid decimal magnitude length " + lenLong + " in xlang body");
    }
    if (lenLong > MAX_MAGNITUDE_BYTES) {
      throw new IllegalArgumentException(
          "Decimal magnitude length exceeds " + MAX_MAGNITUDE_BYTES + " bytes: " + lenLong);
    }
    int len = (int) lenLong;
    buffer.checkReadableBytes(len);
    byte[] magnitudeBytes = buffer.readBytes(len);
    if (magnitudeBytes[len - 1] == 0) {
      throw new IllegalArgumentException("Non-canonical decimal body: trailing zero byte");
    }
    byte[] magnitude = toBigEndian(magnitudeBytes);
    BigInteger abs = new BigInteger(1, magnitude);
    if (abs.signum() == 0) {
      throw new IllegalArgumentException("Big decimal encoding must not represent zero");
    }
    return sign == 0 ? abs : abs.negate();
  }

  private static void checkBinaryBodyLength(int len) {
    if (len <= 0) {
      throw new DeserializationException("Decimal body length must be positive: " + len);
    }
    if (len > MAX_MAGNITUDE_BYTES + 1) {
      throw new DeserializationException(
          "Decimal magnitude exceeds " + MAX_MAGNITUDE_BYTES + " bytes");
    }
  }

  static boolean magnitudeExceedsLimit(MemoryBuffer buffer, int len) {
    int readerIndex = buffer.readerIndex();
    byte first = buffer.getByte(readerIndex);
    // Keep accepting redundant native sign extension. At this length, only a non-sign prefix or
    // the exact negative power -2^(MAX_MAGNITUDE_BITS) has a magnitude above the limit.
    if (first == 0) {
      return false;
    }
    if (first == -1) {
      for (int i = 1; i < len; i++) {
        if (buffer.getByte(readerIndex + i) != 0) {
          return false;
        }
      }
    }
    return true;
  }

  static boolean magnitudeExceedsLimit(BigInteger value) {
    int bitLength = value.bitLength();
    if (bitLength != MAX_MAGNITUDE_BITS) {
      return bitLength > MAX_MAGNITUDE_BITS;
    }
    // BigInteger.bitLength() is one below abs().bitLength() only for negative powers of two.
    // Check that shape solely at the limit so common negative values do not scan their words.
    return value.signum() < 0 && value.getLowestSetBit() == MAX_MAGNITUDE_BITS;
  }

  private static boolean canUseSmallEncoding(BigInteger value) {
    if (value.compareTo(LONG_MIN) < 0 || value.compareTo(LONG_MAX) > 0) {
      return false;
    }
    // The small form reserves the low header bit to distinguish small/big encodings,
    // so the zigzag value itself must still fit in 63 bits before the final << 1.
    long zigZag = encodeZigZag64(value.longValue());
    return (zigZag & Long.MIN_VALUE) == 0;
  }

  private static long encodeZigZag64(long value) {
    return (value << 1) ^ (value >> 63);
  }

  private static long decodeZigZag64(long value) {
    return (value >>> 1) ^ -(value & 1L);
  }

  private static byte[] toCanonicalLittleEndianMagnitude(BigInteger abs) {
    byte[] bigEndian = abs.toByteArray();
    int start = 0;
    while (start < bigEndian.length - 1 && bigEndian[start] == 0) {
      start++;
    }
    int len = bigEndian.length - start;
    if (len <= 0) {
      throw new IllegalArgumentException("Zero must use the small decimal encoding");
    }
    byte[] littleEndian = new byte[len];
    for (int i = 0; i < len; i++) {
      littleEndian[i] = bigEndian[bigEndian.length - 1 - i];
    }
    return littleEndian;
  }

  private static byte[] toBigEndian(byte[] littleEndian) {
    byte[] bigEndian = new byte[littleEndian.length];
    for (int i = 0; i < littleEndian.length; i++) {
      bigEndian[i] = littleEndian[littleEndian.length - 1 - i];
    }
    return bigEndian;
  }
}

File: csharp/src/Fory/DecimalSerializer.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System.Numerics;

namespace Apache.Fory;

public readonly record struct ForyDecimal(BigInteger UnscaledValue, int Scale);

public sealed class DecimalSerializer : Serializer<decimal>
{
    // System.Decimal owns a 96-bit coefficient. Keep its native read bound separate from the
    // wider arbitrary-precision ForyDecimal bound used by compatible scalar conversion.
    private const int MinScale = 0;
    private const int MaxScale = 28;
    private const int MaxMagnitudeBytes = 12;
    private static readonly BigInteger UInt32Mask = uint.MaxValue;

    public override decimal DefaultValue => 0m;

    public override void WriteData(WriteContext context, in decimal value, bool hasGenerics)
    {
        _ = hasGenerics;
        (int scale, BigInteger unscaled) = ToParts(value);
        DecimalCodec.Write(context.Writer, scale, unscaled, maxMagnitudeBytes: null);
    }

    public override decimal ReadData(ReadContext context)
    {
        (int scale, BigInteger unscaled) =
            DecimalCodec.Read(context.Reader, MinScale, MaxScale, MaxMagnitudeBytes);
        return FromParts(scale, unscaled);
    }

    private static (int Scale, BigInteger Unscaled) ToParts(decimal value)
    {
        int[] bits = decimal.GetBits(value);
        BigInteger unscaled =
            ((BigInteger)(uint)bits[2] << 64) |
            ((BigInteger)(uint)bits[1] << 32) |
            (uint)bits[0];
        if ((bits[3] & unchecked((int)0x8000_0000)) != 0)
        {
            unscaled = BigInteger.Negate(unscaled);
        }

        return ((bits[3] >> 16) & 0xFF, unscaled);
    }

    private static decimal FromParts(int scale, BigInteger unscaled)
    {
        if (scale is < 0 or > 28)
        {
            throw new InvalidDataException($"decimal scale {scale} is outside System.Decimal range");
        }

        bool negative = unscaled.Sign < 0;
        BigInteger magnitude = BigInteger.Abs(unscaled);
        if ((magnitude >> 96) != BigInteger.Zero)
        {
            throw new InvalidDataException("decimal magnitude exceeds System.Decimal range");
        }

        int lo = unchecked((int)(uint)(magnitude & UInt32Mask));
        int mid = unchecked((int)(uint)((magnitude >> 32) & UInt32Mask));
        int hi = unchecked((int)(uint)((magnitude >> 64) & UInt32Mask));
        return new decimal(lo, mid, hi, negative, (byte)scale);
    }
}

internal sealed class ForyDecimalSerializer : Serializer<ForyDecimal>
{
    public override ForyDecimal DefaultValue => default;

    public override void WriteData(WriteContext context, in ForyDecimal value, bool hasGenerics)
    {
        _ = hasGenerics;
        if (value.Scale is < DecimalCodec.MinScale or > DecimalCodec.MaxScale)
        {
            throw new InvalidDataException(
                $"decimal scale {value.Scale} is outside range " +
                $"[{DecimalCodec.MinScale}, {DecimalCodec.MaxScale}]");
        }

        DecimalCodec.Write(
            context.Writer,
            value.Scale,
            value.UnscaledValue,
            DecimalCodec.MaxMagnitudeBytes);
    }

    public override ForyDecimal ReadData(ReadContext context)
    {
        (int scale, BigInteger unscaled) =
            DecimalCodec.Read(
                context.Reader,
                DecimalCodec.MinScale,
                DecimalCodec.MaxScale,
                DecimalCodec.MaxMagnitudeBytes);
        return new ForyDecimal(unscaled, scale);
    }
}

internal static class DecimalCodec
{
    public const int MinScale = -10_000;
    public const int MaxScale = 10_000;
    public const int MaxMagnitudeBytes = 10_000;
    private static readonly BigInteger LongMin = long.MinValue;
    private static readonly BigInteger LongMax = long.MaxValue;

    public static void Write(
        ByteWriter buffer,
        int scale,
        BigInteger unscaled,
        int? maxMagnitudeBytes)
    {
        if (CanUseSmallEncoding(unscaled))
        {
            long smallValue = (long)unscaled;
            ulong zigzag = EncodeZigZag64(smallValue);
            buffer.WriteVarInt32(scale);
            buffer.WriteVarUInt64(zigzag << 1);
            return;
        }

        BigInteger magnitude = BigInteger.Abs(unscaled);
        if (magnitude.IsZero)
        {
            throw new InvalidDataException("zero must use the small decimal encoding");
        }

        if (maxMagnitudeBytes is int maxBytes &&
            magnitude.GetBitLength() > (long)maxBytes * 8)
        {
            throw new InvalidDataException(
                $"decimal magnitude exceeds limit {maxBytes}");
        }

        byte[] magnitudeBytes = magnitude.ToByteArray(isUnsigned: true, isBigEndian: false);
        ulong meta = ((ulong)magnitudeBytes.Length << 1) | (unscaled.Sign < 0 ? 1UL : 0UL);
        ulong header = (meta << 1) | 1UL;
        buffer.WriteVarInt32(scale);
        buffer.WriteVarUInt64(header);
        buffer.WriteBytes(magnitudeBytes);
    }

    public static (int Scale, BigInteger Unscaled) Read(
        ByteReader buffer,
        int minScale,
        int maxScale,
        int maxMagnitudeBytes)
    {
        int scale = buffer.ReadVarInt32();
        if (scale < minScale || scale > maxScale)
        {
            throw new InvalidDataException(
                $"decimal scale {scale} is outside range [{minScale}, {maxScale}]");
        }

        ulong header = buffer.ReadVarUInt64();
        if ((header & 1UL) == 0UL)
        {
            return (scale, new BigInteger(DecodeZigZag64(header >> 1)));
        }

        ulong meta = header >> 1;
        ulong lenLong = meta >> 1;
        if (lenLong == 0 || lenLong > int.MaxValue)
        {
            throw new InvalidDataException($"invalid decimal magnitude length {lenLong}");
        }

        if (lenLong > (ulong)maxMagnitudeBytes)
        {
            throw new InvalidDataException(
                $"decimal magnitude length {lenLong} exceeds limit {maxMagnitudeBytes}");
        }

        int length = checked((int)lenLong);
        byte[] magnitudeBytes = buffer.ReadBytes(length);
        if (magnitudeBytes[^1] == 0)
        {
            throw new InvalidDataException("non-canonical decimal magnitude bytes: trailing zero byte");
        }

        BigInteger magnitude = new(magnitudeBytes, isUnsigned: true, isBigEndian: false);
        if (magnitude.IsZero)
        {
            throw new InvalidDataException("big decimal encoding must not represent zero");
        }

        return (scale, (meta & 1UL) == 0UL ? magnitude : BigInteger.Negate(magnitude));
    }

    private static bool CanUseSmallEncoding(BigInteger value)
    {
        if (value < LongMin || value > LongMax)
        {
            return false;
        }

        ulong zigzag = EncodeZigZag64((long)value);
        return (zigzag & (1UL << 63)) == 0;
    }

    private static ulong EncodeZigZag64(long value)
    {
        return unchecked((ulong)((value << 1) ^ (value >> 63)));
    }

    private static long DecodeZigZag64(ulong value)
    {
        return unchecked((long)((value >> 1) ^ (ulong)-(long)(value & 1UL)));
    }
}
